code/discriminative.py

```diff
@@ -0,0 +1,127 @@
+        out = self.linear(sr_emb)
+        return out
+
+    def train(self, emails, all_emails, w2v):
+        all_emails_len = len(all_emails)
+        loss_criteria = nn.CrossEntropyLoss()
+        pos_label = autograd.Variable(torch.LongTensor([1])) #labels for correct mails
+        neg_label = autograd.Variable(torch.LongTensor([0])) #labels for incorrect mails
+        optimizer = optim.RMSprop(self.parameters(), lr=0.0001, alpha=0.99, momentum=0.0)
+        # optimizer = optim.Adam(self.parameters(), lr=0.001, betas=(0.9, 0.999), eps=1e-08, weight_decay=0)
+
+        for epoch in range(self.epochs):
+            print 'running epoch ', epoch
+            start = time.time()
+            epoch_loss = 0.0
+            for i in range(len(emails)):
+                # if i%1000 == 0:
+                #     print 'email processed =', i
+                sender_id = utils.get_userid(emails[i, constants.SENDER_EMAIL])
+                email_word_reps = w2v.get_sentence(emails[i, constants.EMAIL_BODY])
+                # if no word_rep was found for any of the words in the emails, ignore this case
+                if len(email_word_reps) == 0:
+                    continue
+                # gets the average email embedding based on word embeddings of all the words in the mail
+                email_rep = np.mean(email_word_reps, axis=0)
+                recv_list = emails[i, constants.RECEIVER_EMAILS].split('|')
+                for recv in recv_list:
+                    optimizer.zero_grad()
+                    recv_id = utils.get_userid(recv)
+                    # if sender or receiver is not an enron email id, we ignore this data point
+                    if sender_id is None or recv_id is None:
+                        continue
+                    # if valid sender and receiver pairs have been found update their frequencies
+                    self.emailid_train_freq[emails[i, constants.SENDER_EMAIL]] = self.emailid_train_freq.get(
+                        emails[i, constants.SENDER_EMAIL], 0) + 1
+                    self.emailid_train_freq[recv] = self.emailid_train_freq.get(recv, 0) + 1
+                    # do the forward pass
+                    pred_out = self.forward(autograd.Variable(torch.LongTensor([sender_id])),
+                                                  autograd.Variable(torch.LongTensor([recv_id])),
+                                                  autograd.Variable(torch.from_numpy(email_rep)))
+                    
+                    loss = loss_criteria(pred_out,pos_label)
+                    # propagate the loss backward and compute the gradient
+                    loss.backward()
+                    # change weights based on gradient value
+                    optimizer.step()
+
+                    #now train for a negative sample
+                    #get a random email from all_emails where the receiver is not recr, and the mail is not empty
+                    while True:
```

Comment:
Above comment applies here too.
Ideally, we should sample mails once and then train over multiple epochs

Based on how long this takes to run we can reprioritize this.